controllers/project_controller.go

```diff
@@ -0,0 +1,370 @@
+			return ctrl.Result{}, err
+		}
+
+		helper.PushFinalizer(project, internal.FinalizerName)
+
+		project.Status = registriesv1alpha1.ProjectStatus{Phase: registriesv1alpha1.ProjectStatusPhaseReady}
+
+	case registriesv1alpha1.ProjectStatusPhaseReady:
+		err := r.assertExistingProject(ctx, harborClient, project)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+	case registriesv1alpha1.ProjectStatusPhaseTerminating:
+		// Delete the project via harbor API
+		err := r.assertDeletedProject(ctx, reqLogger, harborClient, project)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	return r.updateProjectCR(ctx, harbor, originalProject, project)
+}
+
+func (r *ProjectReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Create a new controller
+	c, err := controller.New("project-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource Project
+	err = c.Watch(&source.Kind{Type: &registriesv1alpha1.Project{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Add a handler to watch for changes in the secondary resource, User
+	watchHandler := &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &registriesv1alpha1.Project{},
+	}
+
+	// Watch for changes to secondary resources
+	err = c.Watch(&source.Kind{Type: &registriesv1alpha1.User{}}, watchHandler)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// updateProjectCR compares the new CR status and finalizers with the pre-existing ones and updates them accordingly.
+func (r *ProjectReconciler) updateProjectCR(ctx context.Context, parentInstance *registriesv1alpha1.Instance, originalProject, project *registriesv1alpha1.Project) (ctrl.Result, error) {
+	if originalProject == nil || project == nil {
+		return ctrl.Result{},
+			fmt.Errorf("cannot update project '%s' because the original project is nil", project.Spec.Name)
+	}
+
+	// Update Status
+	if !reflect.DeepEqual(originalProject.Status, project.Status) {
+		originalProject.Status = project.Status
+		if err := r.Client.Status().Update(ctx, originalProject); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Set owner reference
+	if (originalProject.OwnerReferences == nil ||
+		len(originalProject.OwnerReferences) == 0) &&
+		parentInstance != nil {
+		err := ctrl.SetControllerReference(parentInstance, originalProject, r.Scheme)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Update Finalizer
+	if !reflect.DeepEqual(originalProject.Finalizers, project.Finalizers) {
+		originalProject.SetFinalizers(project.Finalizers)
+	}
+
+	if err := r.Client.Update(ctx, originalProject); err != nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// assertDeletedProject deletes a Harbor project, first ensuring its existence.
+func (r *ProjectReconciler) assertDeletedProject(ctx context.Context, log logr.Logger, harborClient *h.RESTClient,
```

Comment:
I'm not sure what this function is supposed to do. As I understand it, this will always fail. How can we get a deleted project and expect the error to be `nil`?